rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -318,22 +318,19 @@ class Publisher : public PublisherBase
   >
   publish(std::unique_ptr<T, PublishedTypeDeleter> msg)
   {
-    // Avoid allocating when not using intra process.
     if (!intra_process_is_enabled_) {
       // In this case we're not using intra process.
-      ROSMessageType ros_msg;
-      rclcpp::TypeAdapter<MessageT>::convert_to_ros_message(*msg, ros_msg);
-      return this->do_inter_process_publish(ros_msg);
+      auto ros_msg_ptr = std::make_unique<ROSMessageType>();
+      rclcpp::TypeAdapter<MessageT>::convert_to_ros_message(*msg, *ros_msg_ptr);
+      this->do_inter_process_publish(*ros_msg_ptr);
```

Comment:
why not also move the unique_ptr as all other functions do it ?

Reply:
Because [do_inter_process_publish](https://github.com/ros2/rclcpp/blob/b007204fd83828c11f80bf4712d237bfb2431f6e/rclcpp/include/rclcpp/publisher.hpp#L450-L451) takes a constref, not a unique_ptr.

Reply:
My idea was more to change the https://github.com/ros2/rclcpp/blob/b007204fd83828c11f80bf4712d237bfb2431f6e/rclcpp/include/rclcpp/publisher.hpp#L450-L451 to also take a unique_ptr, so that the interface would prevent this type of mistake. I had a further look and found, that it needs to be constref, as the external publish function uses it.